### PR TITLE
 [FLINK-14787][configuration] Add configure method to StreamExecutionEnvironment

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR adds ability to configure StreamExecutionEnvironment from ReadableConfig

It is based on #10222 

## Brief change log

- Added necessary ConfigOptions
- Added possibility to configure StreamExecutionEnvironment from ReadableConfig
- Make the enum matching in Configuration more lenient in regards to case sensitivity (this is needed to match enums that do not follow enum naming convention e.g. `TimeCharacteristic`)
- Regenerated corresponding documentation tables

## Verifying this change

This change added tests:
* StreamExecutionEnvironmentConfigurationTest
* StreamExecutionEnvironmentComplexConfigurationTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no /** don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/cache/DistributedCache.java [hunk 1]
```diff
@@ -79,6 +80,28 @@ public DistributedCacheEntry(String filePath, Boolean isExecutable, byte[] blobK
 			this(filePath, isExecutable, blobKey, false);
 		}
 
+		@Override
```
Comment (kl0u) [inline flink-core/src/main/java/org/apache/flink/api/common/cache/DistributedCache.java]: I must be missing something but are these methods relevant to this PR?

File: flink-core/src/main/java/org/apache/flink/configuration/ReadableConfigToConfigurationAdapter.java [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * A very naive adapter between {@link ReadableConfig} and {@link Configuration}.
+ * It is used to bridge some of the old public interfaces that work with {@link Configuration} even though they
```
Comment (kl0u) [inline flink-core/src/main/java/org/apache/flink/configuration/ReadableConfigToConfigurationAdapter.java]: We can remove the "very naive" and say "simple". "Naive" sounds negative and defensive.

File: flink-core/src/main/java/org/apache/flink/configuration/ReadableConfigToConfigurationAdapter.java [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+ * should actually work with {@link ReadableConfig}.
+ */
+@Internal
+public class ReadableConfigToConfigurationAdapter extends Configuration {
+	private final ReadableConfig backingConfig;
+
+	public ReadableConfigToConfigurationAdapter(ReadableConfig backingConfig) {
+		this.backingConfig = backingConfig;
```
Comment (kl0u) [inline flink-core/src/main/java/org/apache/flink/configuration/ReadableConfigToConfigurationAdapter.java]: It seems that here we can have a `checkNotNull()` for the argument, right?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamPipelineOptions.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+import org.apache.flink.configuration.PipelineOptions;
+import org.apache.flink.configuration.description.Description;
+import org.apache.flink.configuration.description.TextElement;
+import org.apache.flink.streaming.api.TimeCharacteristic;
+
+/**
+ * The {@link ConfigOption} configuration options for job execution. Those are stream specific options.
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamPipelineOptions.java]: `{@link ConfigOption} configuration options` -> `{@link ConfigOption configuration options}` ? I think it seems better.

File: flink-core/src/main/java/org/apache/flink/api/common/cache/DistributedCache.java (hunk 1 above)
Comment (dawidwys) [reply flink-core/src/main/java/org/apache/flink/api/common/cache/DistributedCache.java]: That's not mandatory. I could've implemented it in tests in an asserter. I think it does not harm to have a proper `equals/hashcode`
